modopt/opt/proximity.py

```diff
@@ -434,3 +435,104 @@ def _cost_method(self, *args, **kwargs):
 
         return np.sum([operator.cost(data) for operator, data in
                        zip(self.operators, args[0])])
+
+
+class OrderedWeightedL1Norm(ProximityParent):
+    r"""Ordered Weighted L1 norm proximity operator
+
+    This class defines the OWL proximity operator described in [F2014]
+
+    Parameters
+    ----------
+    weights : np.ndarray
+        Weights values they should be sorted in a non-increasing order
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from modopt.opt.proximity import OrderedWeightedL1Norm
+    >>> A = np.arange(5)*5
+    array([ 0,  5, 10, 15, 20])
+    >>> prox_op = OrderedWeightedL1Norm(np.arange(5))
+    >>> prox_op.weights
+    array([4, 3, 2, 1, 0])
+    >>> prox_op.op(A)
+    array([ 0.,  4.,  8., 12., 16.])
+    >>> prox_op.cost(A, verbose=True)
+     - OWL NORM (X): 150
+    150
+    """
+
+    def __init__(self, weights):
+
+        if any([weights_i < 0 for weights_i in np.squeeze(weights)]):
+            raise ValueError("All the entries of the weights should be"
+                             " positive")
+        self.weights = np.sort(np.squeeze(weights))[::-1]
+        self.op = self._op_method
+        self.cost = self._cost_method
+
+    def _op_method(self, data, extra_factor=1.0):
+        """Operator
+
+        This method returns the input data after the a clustering and a
+        thresholding. Implements (Eq 24) in F2014
+
+        Parameters
+        ----------
+        data : np.ndarray
+            Input data array
+        extra_factor : float
+            Additional multiplication factor
+
+        Returns
+        -------
+        np.ndarray thresholded data
+
+        """
+
+        # Update threshold with extra factor.
+        threshold = self.weights * extra_factor
+
+        # Sorting the absolute value of the input vector
+        data_abs = np.abs(data)
+        idx = np.argsort(np.squeeze(data_abs))[::-1]
+        data_abs = data_abs[idx]
+
+        # Projection onto the monotone non-negative cone using
+        # isotonic_regression
+
+        data_abs = isotonic_regression(data_abs - threshold, y_min=0,
+                                       increasing=False)
+        # Unsorting the data
+        inv_idx = np.zeros_like(idx)
+        inv_idx[idx] = np.arange(len(data))
+        data_abs = data_abs[inv_idx]
```

Comment:
I think it's a bit of a complex method to "unsort" the data, and hard to read.
A simpler way would be:
```python
data_abs = [data_abs[i] for i in idx]
```

Reply:
I am not sure that this is super efficient to do what you proposed. I really don"t see the difficulty here

Reply:
Not even talking about efficiency here since I don't think this part is the bottleneck.
However in terms of readability I don't think this is easy to read.
The first solution I proposed was wrong, this is working:
```python
data_abs_unsorted = np.empty_like(data)
data_abs_unsorted[idx] = data_abs
```
In addition to being in my opinion more readable, it has the advantage of being 2 to 4 times faster, depending on the size of the data (tested in a notebook).

Reply:
Hey more importantly, I have tried @zaccharieramzi 's method from https://github.com/CEA-COSMIC/ModOpt/pull/52#discussion_r281522685 , It does **not** unsort the data.

Reply:
This one works better for me, changing it